orttraining/orttraining/python/training/onnxblock/optim/optim.py

```diff
@@ -211,3 +211,82 @@ def build(self, parameters):
         )
 
         return updated_flag_name
+
+
+class SGDOptimizer(blocks.Block):
+    """Adds a SGDOptimizer node to the onnx model."""
+
+    def __init__(self, learning_rate: float):
+        super().__init__()
+        self._learning_rate = learning_rate
+
+    def build(self, learning_rate_name, weights_name, gradients_name):
+        # get the model to manipulate
+        onnx_model = self.base
+
+        # add the sgd node to the onnx model
+        sgd_input_names = [learning_rate_name, weights_name, gradients_name]
+        sgd_output_name = _graph_utils.generate_graph_name("update_completed")
+        sgd_output_names = [sgd_output_name]
+
+        # add the sgd node to the onnx model
+        sgd_node = onnx.helper.make_node(
+            "SGDOptimizerV2",
+            sgd_input_names,
+            sgd_output_names,
+            name=_graph_utils.generate_graph_name("SGDOptimizerV2"),
+            domain="com.microsoft",
+        )
+        onnx_model.graph.node.append(sgd_node)
+
+        return sgd_output_name
+
+
+class SGD(onnxblock_module.ForwardBlock):
+    """Builds SGD optimizer onnxblock for the given training parameters.
+
+    Creates a block that updates the model parameters based on the calculated
+    gradient following the SGD algorithm.
+
+    Args:
+        learning_rate: float indicating the learning rate.
+    Returns:
+        Returns a string of the output names from this optimizer node.
+    """
+
+    def __init__(self, learning_rate: float):
+        super().__init__()
+        self._sgd = SGDOptimizer(learning_rate)
+
+    def build(self, parameters):
+        """Returns an SGD optimizer model based on the input parameters."""
+
+        # get the model to manipulate and update its namespace
+        onnx_model = self.base
+
+        # TODO: Avoid hard coded input/output strings
+        learning_rate_name = "lr"
```

Comment:
where are these hard coded names used? In the c++ optimizer.cc?

Reply:
here
https://github.com/microsoft/onnxruntime/blob/c424e42594d92daba54f264c1c7409e53529d933/orttraining/orttraining/core/graph/training_op_defs.cc#L1299

Reply:
These are graph input names. They need not match the schema names.

They should match the names expected by the `Optimizer`:

https://github.com/microsoft/onnxruntime/blob/31cbd63af7a8a8582865882b1870719724b6e9b4/orttraining/orttraining/training_api/optimizer.cc#L22

Reply:
@pengwa do you happen to know why `step` is not an input to the SGDOptimizerV2?
From optimizer.cc, it seems to be a required input.

Reply:
I believe SGD don't need step count (for tuning learning rate). Then I think I should not put step in  CommonOptimizerInputs earlier... Only add it when needed (in adamw)